src/dataflow/src/source/kafka.rs

```diff
@@ -100,13 +101,19 @@ impl SourceReader for KafkaSourceReader {
         };
 
         let KafkaSourceConnector {
-            addrs,
+            connector,
             topic,
-            config_options,
             group_id_prefix,
             cluster_id,
             ..
         } = kc;
+        let (addrs, config_options) = match connector {
+            KafkaConnector::Literal(KafkaConnectorLiteral {
+                addrs,
+                config_options,
+            }) => (addrs, config_options),
+            _ => unreachable!("Only Literal is allowed in Dataflow layer"),
```

Comment:
This unreachable here implies that these types aren't sliced quite right. ISTM that you want something like this:

```rust
    pub struct KafkaConnectorLiteral {
        pub addrs: KafkaAddrs,
        // Represents options specified by user when creating the source, e.g.
        // security settings.
        pub config_options: BTreeMap<String, String>,
    }

    pub struct KafkaSourceConfig {
        pub topic: String,
        // Map from partition -> starting offset
        pub start_offsets: HashMap<i32, i64>,
        pub group_id_prefix: Option<String>,
        pub cluster_id: Uuid,
        /// If present, include the timestamp as an output column of the source with the given name
        pub include_timestamp: Option<IncludedColumnPos>,
        /// If present, include the partition as an output column of the source with the given name.
        pub include_partition: Option<IncludedColumnPos>,
        /// If present, include the topic as an output column of the source with the given name.
        pub include_topic: Option<IncludedColumnPos>,
        /// If present, include the offset as an output column of the source with the given name.
        pub include_offset: Option<IncludedColumnPos>,
    }

    pub struct KafkaConnectorAndSourceConfig {
        connector: KafkaConnector,
        config: KafkaSourceConfig,
    }
```

Then catalog::Connector can contain a `KafkaConnector`, `catalog::Source` can contain a `KafkaSourceConfig` and a the ID of a `catalog::Connector`, and the dataflow layer just sees the combined struct `KafkaConnectorAndSourceConfig` and is basically oblivious to the split (for now).

`KafkaConnectorAndSourceConfig` is an awkward name but I don't have anything better atm.

Reply:
I think I am failing to understand something here since it seems like the `KafkaConnectorAndSourceConfig` would still be containing an enum with a variant we want to keep out of the dataflow layer. I agree though that it should be possible to slice the types in a way where the catalog representation can have either a literal or a globalID but the dataflow can only ever receive a literal.

Reply:
I think it might actually work better to try something like 
```rust
pub enum KafkaSource {
    Literal {
        broker: String,
        topic: String,
        key: Option<Vec<Ident>>
    },
    Reference {
        connector: GlobalID,
        topic: String,
        key: Option<Vec<Ident>>
    }
}
```
which can then replace the current `Kafka` variant in `CreateSourceConnector` and then the variants in dataflow-types do not need to exist I don't think.

Reply:
There are unreachables in the revision I just pushed, but they are placeholders since replacing them is dependent in work that will come in the next PR.